packages/amplify-graphql-auth-transformer/src/__tests__/relational-field-redaction.test.ts

```diff
@@ -0,0 +1,263 @@
+  'static groups userPools multiple': '{ allow: groups, groups: ["Admin", "Dev"] }',
+  'static groups oidc single': '{ allow: groups, groups: ["Admin"] provider: oidc }',
+  'static groups oidc multiple': '{ allow: groups, groups: ["Admin", "Dev"] provider: oidc }',
+  'dynamic groups userPools': '{ allow: groups, groupsField: "singleGroup" }',
+  'dynamic groups oidc': '{ allow: groups, groupsField: "singleGroup", provider: oidc }',
+  custom: '{ allow: custom }',
+};
+
+const resolveSchema = (schemaTemplate: string, primaryModelRules: string[], relatedModelRules: string[]): string => {
+  let schema = schemaTemplate;
+  const primaryModelAuthRuleStr = primaryModelRules.map((rule) => authRuleNameToTemplateMap[rule]).join('\n');
+  const relatedModelAuthRuleStr = relatedModelRules.map((rule) => authRuleNameToTemplateMap[rule]).join('\n');
+  schema = schema.replace(/<PRIMARY_MODEL_AUTH_RULES>/g, primaryModelAuthRuleStr);
+  schema = schema.replace(/<RELATED_MODEL_AUTH_RULES>/g, relatedModelAuthRuleStr);
+  return schema;
+};
+
+export const makeTransformationExpectation = (
+  dataSourceStrategies: Record<string, ModelDataSourceStrategy>,
+  schemaTemplate: string,
+): ((strategyName: string, primaryModelRules: string[], relatedModelRules: string[], shouldRedactField: boolean) => void) => {
+  const expectation = (strategyName: string, primaryModelRules: string[], relatedModelRules: string[], shouldRedactField: boolean) => {
+    const validSchema = resolveSchema(schemaTemplate, primaryModelRules, relatedModelRules);
+    const out = testTransform({
+      schema: validSchema,
+      authConfig: authConfigWithAllProviders,
+      transformers: makeTransformers(),
+      dataSourceStrategies: constructDataSourceStrategies(validSchema, dataSourceStrategies[strategyName]),
+    });
+    expect(out).toBeDefined();
+    if (shouldRedactField) {
+      expect(out.resolvers['Primary.relatedMany.auth.1.req.vtl']).toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['Primary.relatedOne.auth.1.req.vtl']).toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['RelatedMany.primary.auth.1.req.vtl']).toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['RelatedOne.primary.auth.1.req.vtl']).toContain(SUBSCRIPTION_PROTECTION);
+    } else {
+      expect(out.resolvers['Primary.relatedMany.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['Primary.relatedOne.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['RelatedMany.primary.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+      expect(out.resolvers['RelatedOne.primary.auth.1.req.vtl']).not.toContain(SUBSCRIPTION_PROTECTION);
+    }
+  };
+  return expectation;
+};
+
+export type TestTableRow = [string, string[], string[], boolean];
+
+/**
+ * Primary auth rules - Related auth rules - Redact relational field
+ */
+const testCases: [string[], string[], boolean][] = [
```

Comment:
These test cases are the main part of this unit test, which defines the different auth rules on both relation models and whether to redact the relational field. These cases are tested below in the dimension of data source types and relation types(field or reference)